llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -2181,6 +2181,123 @@ SmallVector<Range> UnPackOp::getIterationDomain(OpBuilder &builder) {
   return ::getIterationDomain(*this, builder);
 }
 
+SmallVector<Operation *>
+UnPackOp::getTiledImplementation(OpBuilder &builder,
+                                 ArrayRef<OpFoldResult> offsets,
+                                 ArrayRef<OpFoldResult> sizes) {
+  if (!hasTensorSemantics())
```

Comment:
Why do we need to be at the tensor level? It is because the op needs an output tensor?

Reply:
The output of tiled unpack is larger than tiling size because we have to handle incomplete tile. Restricting it at tensor addresses the issue. There are a couple reasons lead me to adding the restriction.

1. The upstream version will be at tensor dialect, so having it works at tensor level SGTM.
2. We'll need a larger buffer to store temp result, and copy the data from temp buffer to destination. We can not reuse the destination buffer because the producing output has more data. Vectorization could potentially address the extra buffer issue. Because everything is vector type, they are going to be stored in registers.
3. In IREE's pipeline, we apply the tiling optimization and the vectorization before bufferization. That makes the world easier. Having it works at tensors is good enough for IREE.

We're able to do tiling at memref level, but that introduces buffer allocation. It requires the users to understand that vectorization could remove the allocation. That's too many details for users. Since there are no needs on IREE side, I restrict it at tensor level for now. I'm happy to extend it if there are needs. I'll add a comment for it!

Reply:
Thanks. Sounds great!